src/monodroid/jni/embedded-assemblies.cc

```diff
@@ -119,120 +106,285 @@ EmbeddedAssemblies::install_preload_hooks ()
+			log_warn (LOG_ASSEMBLY, "typemap: unable to find mapping from Java type '%s' to managed type with token ID %u in module [%s]", java_type_name.get (), java_entry->type_token_id, mono_guid_to_string (module->module_uuid));
+			return nullptr;
+		}
+		type_token_id = java_entry->type_token_id;
 #if defined (DEBUG) || !defined (ANDROID)
-	for (TypeMappingInfo *info = java_to_managed_maps; info != nullptr; info = info->next) {
-		/* log_warn (LOG_DEFAULT, "# jonp: checking file: %s!%s for type '%s'", info->source_apk, info->source_entry, java); */
-		const char *e = reinterpret_cast<const char*> (bsearch (java, info->mapping, static_cast<size_t>(info->num_entries), static_cast<size_t>(info->entry_length), TypeMappingInfo_compare_key));
-		if (e == nullptr)
-			continue;
-		return e + info->value_offset;
 	}
 #endif
-	return find_entry_in_type_map (java, jm_typemap, jm_typemap_header);
+
+	if (module->image == nullptr) {
+		module->image = mono_image_loaded (module->assembly_name);
+		if (module->image == nullptr) {
+			// TODO: load
+			log_error (LOG_ASSEMBLY, "typemap: assembly '%s' not loaded yet!", module->assembly_name);
+		}
+
+		if (module->image == nullptr) {
+			log_error (LOG_ASSEMBLY, "typemap: unable to load assembly '%s' when looking up managed type corresponding to Java type '%s'", module->assembly_name, java_type_name.get ());
+			return nullptr;
+		}
+	}
+
+	log_debug (LOG_ASSEMBLY, "typemap: java type '%s' corresponds to managed token id %u (0x%x)", java_type_name.get (), type_token_id, type_token_id);
+	MonoClass *klass = mono_class_get (module->image, static_cast<uint32_t>(type_token_id));
+	if (klass == nullptr) {
+		log_error (LOG_ASSEMBLY, "typemap: unable to find managed type with token ID %u in assembly '%s', corresponding to Java type '%s'", type_token_id, module->assembly_name, java_type_name.get ());
+		return nullptr;
+	}
+
+	MonoReflectionType *ret = mono_type_get_object (mono_domain_get (), mono_class_get_type (klass));
+	if (XA_UNLIKELY (utils.should_log (LOG_TIMING))) {
+		total_time.mark_end ();
+
+		Timing::info (total_time, "Typemap.java_to_managed: end, total time");
+	}
+
+	return ret;
 }
 
-const char*
-EmbeddedAssemblies::typemap_managed_to_java (const char *managed)
+int
+EmbeddedAssemblies::compare_java_name (const char *java_name, const TypeMapJava *entry)
 {
-#if defined (DEBUG) || !defined (ANDROID)
-	for (TypeMappingInfo *info = managed_to_java_maps; info != nullptr; info = info->next) {
-		/* log_warn (LOG_DEFAULT, "# jonp: checking file: %s!%s for type '%s'", info->source_apk, info->source_entry, managed); */
-		const char *e = reinterpret_cast <const char*> (bsearch (managed, info->mapping, static_cast<size_t>(info->num_entries), static_cast<size_t>(info->entry_length), TypeMappingInfo_compare_key));
-		if (e == nullptr)
-			continue;
-		return e + info->value_offset;
+	// Looking for cause of a segfault which occurs only on the build bots, the log entries will be removed later on
```

Comment:
Is this comment still relevant?  I don't see any "log entries" near/after this comment.